ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+- Some concrete error class placement might be confusing and will be clarified later in the process
+
+
+### Converting API Failures
+
+Similar to other SDKs we will decode `Failure` proto messages to native error classes. All `cause`s
+will be decoded recursively and the result will form a hierarchy identical to the one in the proto
+response. These error classes will all be subclasses of an `APIError` and will have names matching
+the original protobuf messages (while replacing `Failure` with `Error` for consistency within Ruby):
+
+```ruby
+module Temporal
+  class ApplicationError < APIError
+  class TimeoutError < APIError
+  class CanceledError < APIError
+  class TerminatedError < APIError
+  class ServerError < APIError
+  class ResetWorkflowError < APIError
+  class ActivityError < APIError
+  class ChildWorkflowExecutionError < APIError
+end
+```
+
+At this point we are not going to be converting user errors back to their original classes. These
+will be decoded into `ApplicationError`s while keeping the original class name, message, stack trace
+and other details. Here's an example of handling an activity error:
+
+```ruby
+class ProcessPaymentActivity < Temporal::Activity
+  class InsufficientFunds < Temporal::NonRetryableError; end
+  class PaymentNetworkError < Temporal::RetryableError
+    attr_reader :status
+
+    def initialize(message, status)
+      super(message)
+      @status = status
+    end
+  end
+
+  def execute(account, amount)
+    if amount < balance_of(account.number)
+      raise InsufficientFunds, "Account #{account.number} does not have enough funds"
+    end
+
+    response = process_payment(account.number, amount)
+    if response.failure?
+      raise PaymentNetworkError, response.message, response.status
+    end
+
+    ...
+  end
+end
+
+class RenewSubscriptionWorkflow < Temporal::Workflow
+  def execute
+    ...
+
+    begin
+      workflow.execute_activity(ProcessPaymentActivity, account, amount)
+    rescue Temporal::ActivityError => error
+      if error.cause.is_a?(Temporal::ApplicationError)
+        case error.cause.original_class
+        when 'ProcessPaymentActivity::InsufficientFunds'
+          # handle insufficient funds error here
+        when 'ProcessPaymentActivity::PaymentNetworkError'
+          # handle payment network error
+          # notice that status is not available
+        end
+      end
+
+      # You'll need to re-raise the error in case it wasn't handled
+      raise
+    end
+
+    ...
+  end
+end
+```
+
+Application errors will be encoded/decoded using the same Payload converters as the other payloads.
+This also means that if an error class implements `#to_json` / `.json_create` (as suggested by the
+[Ruby's JSON module](https://ruby-doc.org/stdlib-3.1.2/libdoc/json/rdoc/JSON.html#module-JSON-label-Custom+JSON+Additions))
+then we will be able to recreate the original error class:
```

Comment:
Where are you deserializing this from? The first error detail payload and ignore the rest?

It can be confusing for a user to know that half the time you deserialize into a known type and half the time you don't based on whether the class supports it. To clarify, you are going to always try to instantiate a class of the app error's "type" and then always try to deserialize JSON and on any error you'll leave it as app error? As a user, am I expected to handle both cases just in case the potentially-separate implementer of the error failed JSON serialization?

We don't support this in any SDK today, can we not support this in Ruby either until we have a holistic solution here? All SDKs suffer this issue, I don't think we should solve now.

Reply:
I'd punt on this but I agree with the general direction.

Reply:
Sounds good, I can see how this is a bigger discussion across all the SDKs. I'll leave this part off for now 👍